src/_pytest/terminal.py

```diff
@@ -803,31 +806,36 @@ def getreports(self, name):
         return values
 
     def summary_warnings(self):
-        if self.hasopt("w"):
-            all_warnings = self.stats.get(
-                "warnings"
-            )  # type: Optional[List[WarningReport]]
-            if not all_warnings:
-                return
+        if not self.hasopt("w"):
+            return
 
-            final = hasattr(self, "_already_displayed_warnings")
-            if final:
-                warning_reports = all_warnings[self._already_displayed_warnings :]
-            else:
-                warning_reports = all_warnings
-            self._already_displayed_warnings = len(warning_reports)
-            if not warning_reports:
-                return
+        all_warnings = self.stats.get("warnings")  # type: Optional[List[WarningReport]]
+        if not all_warnings:
+            return
 
-            reports_grouped_by_message = (
-                collections.OrderedDict()
-            )  # type: collections.OrderedDict[str, List[WarningReport]]
-            for wr in warning_reports:
-                reports_grouped_by_message.setdefault(wr.message, []).append(wr)
+        final = hasattr(self, "_already_displayed_warnings")
+        if final:
+            warning_reports = all_warnings[self._already_displayed_warnings :]
+        else:
+            warning_reports = all_warnings
+        self._already_displayed_warnings = len(warning_reports)
+        if not warning_reports:
+            return
 
-            title = "warnings summary (final)" if final else "warnings summary"
+        grouped = (
+            collections.OrderedDict()
+        )  # type: collections.OrderedDict[str, List[WarningReport]]
+        for wr in warning_reports:
+            if wr.when not in grouped:
+                grouped[wr.when] = collections.OrderedDict()
+            grouped[wr.when].setdefault(wr.message, []).append(wr)
+
+        for when, grouped_by_message in grouped.items():
+            title = (
+                "warnings summary (final)" if final else "warnings summary"
+            ) + " [{}]".format(when)
```

Comment:
I suspect most warnings come from runtest... how about only showing the suffix in case we have more than one category?

Reply:
Also I believe users might get confuse if they see only `warnings summary [config]` or `warnings summary [collect]` (while "runtest" will probably be OK).

How about adding some word to clarify, perhaps `warnings summary [captured during: config]`.

Reply:
@Zac-HD what do you suggest?

I do not thinkg "captured during: config" is much better than "config", is it?

Reply:
IMO it probably *is* better, because few users have a good mental model of pytest's operation and so being very descriptive is useful - most people don't read docs until they have a problem, so the message is our first chance to teach them. 

"Summary of warnings emitted during config stage" would be good for that, for example.

(note that I'm happy to merge anything better than the status quo though!)

Reply:
@Zac-HD 
Ok, understood.
OTOH headings should be short (to allow for more exclamation marks ;)).
The status quo for myself is that I have it like it is here already, and I am happy with it.
So I suggest somebody should pick it up from here, and maybe improve it with regard to what you and @nicoddemus are suggesting.  I'm not gonna do it myself.
Also given that there are conflicts it needs some work anyway (which I am also not doing anytime soon myself).
Therefore, either adopt it with improved headings, or just fix the conflicts, or close it.

Reply:
OK, let's merge as-is then once someone fixes merge conflicts. 

(I'll try to get to it eventually but have some much more urgent projects at the moment)

Reply:
> Therefore, either adopt it with improved headings, or just fix the conflicts, or close it.

To avoid having this lingering here, let's close it then. @Zac-HD feel free to re-open it if you want to get back to this eventually, and thanks!